src/newTest/java/org/opensearch/test/framework/cluster/LocalOpenSearchCluster.java

```diff
@@ -0,0 +1,610 @@
+                masterNodeHttpPorts);
+
+        SortedSet<Integer> nonMasterNodeTransportPorts = PortAllocator.TCP.allocate(clusterName, nonMasterNodeCount, 5000 + forkNumber * 1000 + 310);
+        SortedSet<Integer> nonMasterNodeHttpPorts = PortAllocator.TCP.allocate(clusterName, nonMasterNodeCount, 5000 + forkNumber * 1000 + 210);
+
+        CompletableFuture<Void> nonMasterNodeFuture = startNodes(clusterConfiguration.getNonMasterNodeSettings(), nonMasterNodeTransportPorts,
+                nonMasterNodeHttpPorts);
+
+        CompletableFuture.allOf(masterNodeFuture, nonMasterNodeFuture).join();
+
+        if (isNodeFailedWithPortCollision()) {
+            log.info("Detected port collision for master node. Retrying.");
+
+            retry();
+            return;
+        }
+
+        log.info("Startup finished. Waiting for GREEN");
+
+        waitForCluster(ClusterHealthStatus.GREEN, TimeValue.timeValueSeconds(10), allNodes.size());
+        putDefaultTemplate();
+
+        log.info("Started: {}", this);
+
+    }
+
+    public String getClusterName() {
+        return clusterName;
+    }
+
+    public boolean isStarted() {
+        return started;
+    }
+
+    public void stop() {
+
+        for (Node node : clientNodes) {
+            node.stop();
+        }
+
+        for (Node node : dataNodes) {
+            node.stop();
+        }
+
+        for (Node node : masterNodes) {
+            node.stop();
+        }
+    }
+
+    public void destroy() {
+        stop();
+        clientNodes.clear();
+        dataNodes.clear();
+        masterNodes.clear();
+
+        try {
+            FileUtils.deleteDirectory(clusterHomeDir);
+        } catch (IOException e) {
+            log.warn("Error while deleting " + clusterHomeDir, e);
+        }
+    }
+
+    public Node clientNode() {
+        return findRunningNode(clientNodes, dataNodes, masterNodes);
+    }
+    
+    public Node randomClientNode() {
+        return randomRunningNode(clientNodes, dataNodes, masterNodes);
+    }
+
+    public Node masterNode() {
+        return findRunningNode(masterNodes);
+    }
+
+    public List<Node> getAllNodes() {
+        return Collections.unmodifiableList(allNodes);
+    }
+
+    public Node getNodeByName(String name) {
+        return allNodes.stream().filter(node -> node.getNodeName().equals(name)).findAny().orElseThrow(() -> new RuntimeException(
+                "No such node with name: " + name + "; available: " + allNodes.stream().map(Node::getNodeName).collect(Collectors.toList())));
+    }
+    
+    private boolean isNodeFailedWithPortCollision() {
+        return allNodes.stream().anyMatch(Node::isPortCollision);
+    }
+
+    private void retry() throws Exception {
+        retry++;
+
+        if (retry > 10) {
+            throw new RuntimeException("Detected port collisions for master node. Giving up.");
+        }
+
+        stop();
+
+        this.allNodes.clear();
```

Comment:
This seems like a great place to generate bugs, IMO it would be better to have a `StartingCluster` that is separate and when everything is 'good' then create the actual cluster instance that is handled back to the test cases

Reply:
What bugs are you thinking of? 

Actually, instances of this class are never visible to the test cases. The test cases only see `TestCluster`, which encapsulates this class.

Reply:
Since the cluster instance is stateful while retries are in progress assumptions could be made about what the nodes are, while the collection is cleared this doesn't mean references to the nodes are cleared.

This is more a code smell kind of issue than a specific feedback / direction.  What do you think?

Reply:
Usually, the cluster will be created as a JUnit ExternalResource, which means that the initialization process is completely opaque for the test code itself.  Thus, there are no entities - except the cluster itself - which could observe the state.

Maybe just clarifying this as a comment would be sufficient?

Reply:
Altering this functionality would require a considerable refactor, at this point, things have stabilized and this isn't a dealbreaker, let's continue with this implementation.